src/Features/Core/Diagnostics/AbstractHostDiagnosticUpdateSource.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal abstract class AbstractHostDiagnosticUpdateSource : IDiagnosticUpdateSource
+    {
+        private static ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<DiagnosticData>> _analyzerHostDiagnosticsMap =
+            ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<DiagnosticData>>.Empty;
+
+        protected AbstractHostDiagnosticUpdateSource()
+        {
+            // Register for exception diagnostics from workspace's analyzer manager.
+            WorkspaceAnalyzerManager.AnalyzerExceptionDiagnostic += OnAnalyzerExceptionDiagnostic;
```

Comment:
So the args for exception diagnostic event from compiler layer just provides faulted analyzer + diagnostic. We need to wrap this event with workspace/project context in which analyzer was executed. So I have this wrapping done in WorkspaceAnalyzerManager and then the driver registers for the event before running analyzer, also appending project information whose analysis caused analyzer to throw.
The event can as well be instance field on the analyzer manager, but then the WorkspaceAnalyzerManager will need to passed down into constructor of DiagnosticAnalyzerDriver. Seemed like too much trouble when static event is sufficient. In fact, I can even move the WorkspaceAnalyzerManager.AnalyzerExceptionDiagnostic directly into AbstractHostDiagnosticUpdateSource.